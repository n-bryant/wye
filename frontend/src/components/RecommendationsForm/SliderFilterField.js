import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Slider from "@material-ui/core/Slider";
import Typography from "@material-ui/core/Typography";

import styles from "./SliderFilterField.styles";

/**
 * renders a field for electing a boolean value in the form of a Switch
 */
export const SliderFilterField = props => {
  const classnames = SliderFilterField.classnames(props);
  const {
    classes,
    name,
    min,
    max,
    minType,
    maxType,
    label,
    setFieldValue,
    value,
    labelFormatHandler
  } = props;
  const [sliderValue, setSliderValue] = React.useState([min, max]);

  /**
   * handle changes to the Slider,
   * and update formik field values
   * @param {String} option
   */
  const handleChange = (_event, newValue) => {
    setSliderValue(newValue);
    setFieldValue(name, {
      ...value,
      gameFilters: {
        ...value.gameFilters,
        [minType]: newValue[0],
        [maxType]: newValue[1]
      }
    });
  };

  return (
    <div className={classnames.root()}>
      <Typography
        id="range-filter-slider"
        className={classnames.element("label")}
        variant="body2"
        gutterBottom={true}
      >
        {label}
      </Typography>
      <div className={classnames.element("valueLabelContainer")}>
        <Typography variant="body2">
          {labelFormatHandler(sliderValue[0])}
        </Typography>
        <Typography variant="body2">
          {labelFormatHandler(sliderValue[1])}
        </Typography>
      </div>
      <div>
        <Slider
          classes={{
            root: classes.sliderRoot,
            thumb: classes.thumb,
            track: classes.track,
            rail: classes.rail
          }}
          min={min}
          max={max}
          value={sliderValue}
          onChange={handleChange}
          valueLabelDisplay="off"
          aria-labelledby="range-filter-slider"
          getAriaValueText={() => `${sliderValue}`}
        />
      </div>
    </div>
  );
};
SliderFilterField.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}SliderFilterField`
);
SliderFilterField.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    label: PropTypes.string,
    sliderRoot: PropTypes.string,
    thumb: PropTypes.string,
    track: PropTypes.string,
    rail: PropTypes.string,
    valueLabelContainer: PropTypes.string
  }),
  // Formik field name
  name: PropTypes.string,
  // the label to display for the field
  label: PropTypes.string,
  // the minimum allowed value
  min: PropTypes.number.isRequired,
  // the maximum allowed value
  max: PropTypes.number.isRequired,
  // the type of the min value
  minType: PropTypes.oneOf([
    "userRating_gte",
    "ownersMin_gte",
    "ownersMax_gte",
    "discount_gte",
    "finalPrice_gte"
  ]).isRequired,
  // the type of the max value
  maxType: PropTypes.oneOf([
    "userRating_lte",
    "ownersMin_lte",
    "ownersMax_lte",
    "discount_lte",
    "finalPrice_lte"
  ]).isRequired,
  // playerFilters field value from Formik form
  value: PropTypes.object,
  // handler for updating users field value
  setFieldValue: PropTypes.func,
  // how to format the value labels
  labelFormatHandler: PropTypes.func.isRequired
};
SliderFilterField.defaultProps = {
  classes: {}
};

// apply styles
export default withStyles(styles)(SliderFilterField);
