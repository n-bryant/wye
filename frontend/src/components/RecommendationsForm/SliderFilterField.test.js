import { Slider } from "@material-ui/core";
import { SliderFilterField } from "./SliderFilterField";

describe("SliderFilterField", () => {
  const props = {
    classes: {
      root: "root",
      label: "label",
      sliderRoot: "sliderRoot",
      thumb: "thumb",
      track: "track",
      rail: "rail",
      valueLabelContainer: "valueLabelContainer"
    },
    name: "name",
    label: "label",
    min: 0,
    max: 100,
    minType: "discount_lte",
    maxType: "discount_gte",
    value: {
      gameFilters: {}
    },
    setFieldValue: jest.fn(),
    labelFormatHandler: jest.fn(x => `${x}%`)
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<SliderFilterField {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should update the formik field value when a change occurs", () => {
    wrapper.find(Slider).prop("onChange")({}, [25, 75]);
    expect(props.setFieldValue).toHaveBeenCalledWith(props.name, {
      gameFilters: {
        [props.minType]: 25,
        [props.maxType]: 75
      }
    });
  });
});
