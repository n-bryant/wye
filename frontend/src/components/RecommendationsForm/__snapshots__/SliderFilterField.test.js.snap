// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderFilterField should render successfully 1`] = `
<div
  className="wye-SliderFilterField root"
>
  <WithStyles(ForwardRef(Typography))
    className="wye-SliderFilterField__label label"
    gutterBottom={true}
    id="range-filter-slider"
    variant="body2"
  >
    label
  </WithStyles(ForwardRef(Typography))>
  <div
    className="wye-SliderFilterField__valueLabelContainer valueLabelContainer"
  >
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      0%
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      variant="body2"
    >
      100%
    </WithStyles(ForwardRef(Typography))>
  </div>
  <div>
    <WithStyles(ForwardRef(Slider))
      aria-labelledby="range-filter-slider"
      classes={
        Object {
          "rail": "rail",
          "root": "sliderRoot",
          "thumb": "thumb",
          "track": "track",
        }
      }
      getAriaValueText={[Function]}
      max={100}
      min={0}
      onChange={[Function]}
      value={
        Array [
          0,
          100,
        ]
      }
      valueLabelDisplay="off"
    />
  </div>
</div>
`;
